File: napkin/src/logpanel/logpanel.cpp
#include "logpanel.h"

#include "../appcontext.h"


LogPanel::LogPanel(QWidget *parent) :
	QDockWidget(parent)
{
	ui.setupUi(this);

	mLogModel = new QStandardItemModel(this);
	ui.treeView->setModel(mLogModel);

	mHeaderLabels.append("Level");
	mHeaderLabels.append("LogMessage");
	mLogModel->setHorizontalHeaderLabels(mHeaderLabels);
    nap::Logger::instance().log.connect(onLogSlot);
    nap::Logger::debug("LogPanel connected");
}

void LogPanel::onMessageReceived(nap::LogMessage msg)
{
    static QMap<nap::LogLevel, LogLevelStyle*> mStyleMap;

    auto levelItem = new QStandardItem(QString::fromStdString(msg.level().name()));
    auto icons = AppContext::get().iconStore();

    LogLevelStyle* style = nullptr;
    if (msg.level() == nap::Logger::debugLevel()) {
        style = &mLogStyleDebug;
    } else if (msg.level() == nap::Logger::infoLevel()) {
        style = &mLogStyleInfo;
    } else if (msg.level() == nap::Logger::warnLevel()) {
        style = &mLogStyleWarn;
    } else if (msg.level() == nap::Logger::fatalLevel()) {
        style = &mLogStyleFatal;
    } else {
        assert(false); // unsupported log level
    }

    levelItem->setBackground(style->color());
    levelItem->setIcon(*style->icon());

    QString message = QString::fromStdString(msg.text());
    auto messageItem = new QStandardItem(message);
    messageItem->setToolTip(message);
    messageItem->setBackground(levelItem->background());

    QList<QStandardItem*> items;
	items << levelItem;
	items << messageItem;
	mLogModel->appendRow(items);
	ui.treeView->scrollTo(mLogModel->index(mLogModel->rowCount() - 1, 0));
}

 

File: py/logpanel.py
from PyQt5.QtGui import *
from PyQt5.QtCore import *
from PyQt5.QtWidgets import *


class LogModel(QStandardItemModel):

    MAX_MESSAGES = 1000

    def __init__(self, ctx):
        """
        @type ctx: appcontext.AppContext
        """
        self.__ctx = ctx
        super(LogModel, self).__init__()
        self.setHorizontalHeaderLabels(['Level', 'Message'])
        self.__ctx.logMessageReceived.connect(self.__onLogMessageReceived)

    def __onLogMessageReceived(self, lvl, lvlName, text):
        while self.rowCount() > self.MAX_MESSAGES:
            self.removeRow(0)
        self.appendRow([
            QStandardItem(lvlName),
            QStandardItem(text)
        ])


class LogPanel(QWidget):
    def __init__(self, ctx):
        """
        @type ctx: appcontext.AppContext
        """
        self.__ctx = ctx
        super(LogPanel, self).__init__()
        self.__logModel = LogModel(ctx)
        self.__treeView = QTreeView()
        self.__treeView.setModel(self.__logModel)
        self.setLayout(QVBoxLayout())
        self.layout().addWidget(self.__treeView)

        self.__ctx.logMessageReceived.connect(self.__onLogMessageReceived)

    def __onLogMessageReceived(self, lvl, lvlName, text):
        idx = self.__logModel.index(self.__logModel.rowCount() -1, 0)
        self.__treeView.scrollTo(idx)